tokio-util/src/sync/task_tracker.rs

```diff
@@ -0,0 +1,591 @@
+        inner: Option<&'a TaskTrackerInner>,
+    }
+}
+
+impl TaskTrackerInner {
+    #[inline]
+    fn new() -> Self {
+        Self {
+            state: AtomicUsize::new(0),
+            on_last_exit: Notify::new(),
+        }
+    }
+
+    #[inline]
+    fn is_done(&self) -> bool {
+        // If empty and closed bit set, then we are done.
+        //
+        // The acquire load will synchronize with the release store of any previous call to
+        // `set_closed` and `drop_task`.
+        self.state.load(Ordering::Acquire) == 1
+    }
+
+    #[inline]
+    fn set_closed(&self) -> bool {
+        // The AcqRel ordering makes the closed bit behave like a `Mutex<bool>` for synchronization
+        // purposes. We do this because it makes the return value of `TaskTracker::{close,reopen}`
+        // more meaningful for the user. Without these orderings, this assert could fail:
+        // ```
+        // // thread 1
+        // some_other_atomic.store(true, Relaxed);
+        // tracker.close();
+        //
+        // // thread 2
+        // if tracker.reopen() {
+        //     assert!(some_other_atomic.load(Relaxed));
+        // }
+        // ```
+        // However, with the AcqRel ordering, we establish a happens-before relationship from the
+        // call to `close` and the later call to `reopen` that returned true.
+        let state = self.state.fetch_or(1, Ordering::AcqRel);
+
+        // If there are no tasks, and if it was not already closed:
+        if state == 0 {
+            self.notify_now();
+        }
+
+        (state & 1) == 0
+    }
+
+    #[inline]
+    fn set_open(&self) -> bool {
+        // See `set_closed` regarding the AcqRel ordering.
+        let state = self.state.fetch_and(!1, Ordering::AcqRel);
+        (state & 1) == 1
+    }
+
+    #[inline]
+    fn add_task(&self) {
+        self.state.fetch_add(2, Ordering::Relaxed);
+    }
+
+    #[inline]
+    fn drop_task(&self) {
+        let state = self.state.fetch_sub(2, Ordering::Release);
+
+        // If this was the last task and we are closed:
+        if state == 3 {
+            self.notify_now();
+        }
+    }
+
+    #[cold]
+    fn notify_now(&self) {
+        // Insert an acquire fence. This matters for `drop_task` but doesn't matter for
+        // `set_closed` since it already uses AcqRel.
+        //
+        // This synchronizes with the release store of any other call to `drop_task`, and with the
+        // release store in the call to `set_closed`. That ensures that everything that happened
+        // before those other calls to `drop_task` or `set_closed` will be visible after this load,
+        // and those things will also be visible to anything woken by the call to `notify_waiters`.
+        self.state.load(Ordering::Acquire);
+
+        self.on_last_exit.notify_waiters();
+    }
+}
+
+impl TaskTracker {
+    /// Create a new `TaskTracker`.
+    ///
+    /// The `TaskTracker` will start out as open.
+    pub fn new() -> Self {
```

Comment:
nit: this constructor should probably have
```suggestion
    #[must_use]
    pub fn new() -> Self {
```